react/lesson1.md

```diff
@@ -36,175 +91,431 @@ cd my-app/
+
+Since it's not exactly HTML, there are some gotchas, of which you'll need to be aware of:
+
+- event handlers are capitalized
+
+HTML: `<button onclick="myFunction()"/>` 
+JSX: `button onClick="myFunction()"/>`
+
+- `className` instead of `class`
 
-export default App;
+HTML: `<button class="myButton"/>`
+JSX: `<button className="myButton"/>`
 
+- must always have one child element
+
+Won't work:
+
+```javascript
+return (
+  <p> Hello! </p>
+  <p> This is my blog. </p>
+);
 ```
 
-Child or `presentational` components are essentially smaller customisable to the data that they receive but usually do not have the data in themselves. These are often referred to as `dumb` components - because they do not process the data in themselves:
+Will work:
 
-```jsx
-import React, { Component } from 'react';
+```javascript
+return (
+  <div>
+    <p> Hello! </p>
+    <p> This is my blog. </p>
+  </div>
+);
+```
 
-class Organisation extends Component {
-  render() {
-    const name = this.props.name;
-    const website = this.props.website;
-    return (
-        <div>
-            Name: {name} <br />
-            Website: {website}
-        </div>
-    );
+## Thinking in React
+
+Let's go over some examples and discuss how we would break everything up into smart and dumb components.
+
+# Event handlers
+
+Time to handle a few user events. We'll add a "difficulty setting" input. For now, whenever the user changes that input, we will `console.log` the result.
+
+```javascript
+
+difficultyChange = (event) => {
+  console.log(event.target.value);
+};
+
+//inside render()
+<input type="number" onChange={this.difficultyChange}/>
+```
+
+3 things to notice:
+
+- the curly brackets after `onChange`: surrounding something by two curly brackets is a way to use JavaScript inside JSX
+- `difficultyChange` is written as an arrow function: this is a good way to avoid issues with the `this` context in React. You can read more about it here: https://facebook.github.io/react/docs/handling-events.html
+- the `event` object: this is a special object passed to all event handlers in React. In our case we can use it to access the target DOM element of that event, and more specifically, its value.
+
+# JSX: Map Arrays to Components
+
+Because works so well with JavaScript, we can even it in for loops and build JS arrays that contain JSX elements.
+
+Adding more game buttons would be easy, we can just copy-paste them one after another. But what if need a lot of them? Or what if we want to make their number dynamic, chosen by the user?
+
+Let's create a function that returns us an array of game buttons:
+
+```javascript
+buildGame() {
+  let buttons =  [];
+  for(var i=0; i<10; i++) {
+    buttons.push(<GameButton />)
   }
+  return buttons;
 }
+```
 
-export default Organisation;
+And then call it in our `render()` function:
 
+```javascript
+<p className="App-intro">
+  <input type="number" onChange={this.difficultyChange}/>
+  <br/>
+  {this.buildGame()}
+</p>
 ```
```

Comment:
Can we include an example with map here? It's mentioned in the heading for the section and it's often much more common to use `this.props.someField.map` over a `for (let i = 0; i < this.props.someField.length; i++)` in React.

Be aware that we're also going to get a warning about not having a `key` on each element here.